README.md
# Postcard Creator API [![Build Status](https://travis-ci.org/gido/postcardcreator.svg?branch=master)](https://travis-ci.org/gido/postcardcreator)
A node.js API for the [Swiss Post Postcard Creator](http://postcardcreator.post.ch).

# Basic usage

```javascript
var path = require('path'),
    Postcardcreator = require('postcardcreator'),
    SSOHelper = require('postcardcreator/helper/SSOPostHelper')(),
    Postcard = Postcardcreator.Postcard;

SSOHelper.getPostcardcreatorToken(postcard_user, postcard_pass, function(err, data) {

    if (err) {
        console.error(err);
        return;
    }

    var client = new Postcardcreator(data.token);

    var message = "Hello, here is a picture of me. Best!";
    var assetStream = fs.createReadStream(path.join(__dirname, 'me_under_the_sun.jpg'));
    // here is my real address you can use it to send me picture of your works ;-)
    var recipient = {
        salutation: "Monsieur",
        givenName: "Gilles",
        familyName: "Doge",
        company: "Antistatique.net",
        street: "Rue de Sébeillon 9b",
        postCode: "1004",
        place: "Lausanne"
    };

    var postcard = new Postcard(assetStream, message, recipient);

    client.sendPostcard(postcard, function(err, result) {
        if (err) {
            console.log("Error when sending the postcard. ", err);
            handleError(err);
            return;
        }

        console.log("Postcard sent with success !");
        console.log(result);
    });
});
```


# License
MIT

# Related projects
 - [abertschi/postcard_creator_wrapper](https://github.com/abertschi/postcard_creator_wrapper) - Python wrapper around the Rest API of the Swiss Postcard Creator


CHANGELOG.md
CHANGELOG
---------

## 2.2.0
    - Add support to 2.2 PostcardCreator REST endpoint (2.0 is Gone).
    - Better new lines handling

## 2.0.0
    - Add support to the new 2.0 PostcardCreator REST endpoint

## 0.1.0
    - BC: API change of the `Client.sendPostcard` and introduction of a `Postcard` object.
      This new object is responsible of the layout rendering. Can be overrided to change the default template.

## 0.0.2
    - Fix error when no credentials is given
    - added support to register new user
    - more tests

## 0.0.1
    - Initial version


spec/server.js
var express = require('express'),
    bodyParser = require('body-parser'),
    http = require('http'),
    util = require('util'),
    path = require('path'),
    fs = require('fs'),
    multiparty = require('multiparty');

function createTestServer(options) {
    options = options || {};

    var validAuthToken = options.validAuthToken || "ABC123456",
        host = options.host || "localhost",
        port = options.port || 3000;

    var testUserId    = options.testUserId    || 12345;
    var testNewUserId = options.testNewUserId || 67890;
    var testMailingId = options.testMailingId || 30003;
    var testAssetId   = options.testAssetId   || 40001;
    var testOrderId   = options.testOrderId   || 50001;


    // create a local Express App
    var app = express();
    var auth = function (req, res, next) {
        var bearer = req.get('Authorization');
        var token;

        if (bearer) {
          token = bearer.replace('Bearer ', '');
        }

        if (validAuthToken === token) {
          next();
        } else {
          res.statusCode = 401;
          res.statusText = 'Unauthorized';
          res.json({
            "error": "invalid_token",
            "error_description": "Cannot convert access token to JSON"
          }).send();
        }
    };

    app.use(auth);

    app.use(bodyParser.json()); // for parsing application/json
    app.use(bodyParser.urlencoded({ extended: true })); // for parsing application/x-www-form-urlencoded
    app.use(bodyParser.raw({ type: 'image/svg+xml' }));

    app.get('/rest/2.2/users/current', function(req, res) {
        var data = {
            "tenantId": "CHE",
            "userId": testUserId,
            "email": req.user,
            "company": null,
            "companyAdditional": null,
            "sex": "MALE",
            "title": null,
            "givenName": "Bob",
            "familyName": "",
            "address": "My Street",
            "addressAdditional": null,
            "poBox": null,
            "postCode": "1004",
            "place": "Lausanne",
            "country": "CHE",
            "phone": null,
            "language": "fr",
            "customerNo": null,
            "newsletterSubscribed": true,
            "gtcAccepted": true
        };

        res.json(data);
    });

    app.post('/rest/2.2/users', function(req, res) {
        var userUrl = req.protocol + '://' + req.get('host') + '/rest/2.2/users/' + testNewUserId;

        var requiredKeys = [
            "tenantId",
            "email",
            "password",
            "company",
            "sex",
            "givenName",
            "familyName",
            "language",
            "newsletterSubscribed",
            "gtcAccepted",
            "address",
            "postCode",
            "place",
            "country"
        ];

        var data = req.body;

        for (var requiredKey in requiredKeys) {
            if (!data.hasOwnProperty(requiredKey)) {
                res.status(400).end('Error, missing value '+requiredKey);
                return;
            }
        }

        res.status(201).location(userUrl).end();

    });

    app.get('/rest/2.2/users/:user_id/quota', function(req, res) {
        var quota = {
            "quota": 1,
            "globalQuotaExceeded": false
        };

        res.json(quota);
    });

    app.post('/rest/2.2/users/:user_id/mailings', function(req, res) {
        var mailingUrl = req.protocol + '://' + req.get('host') + '/rest/2.2/users/' + req.params.user_id + '/mailings/' + testMailingId;

        // {"name":"Mobile App mailing 2015-04-22 20:03", "productId":2, "source":"MOBILE", "addressFormat":"PERSON_FIRST"}
        if (!req.body || !req.body.name || !req.body.productId || !req.body.source || !req.body.addressFormat) {
            return res.status(400)
                .send('Bad Request, missing JSON POST parameters')
                .end();

        }

        // all is fine, return a 201 Created.
        res.status(201)
            .location(mailingUrl)
            .end();
    });

    app.post('/rest/2.2/users/:user_id/assets', function(req, res) {
        var assetUrl = req.protocol + '://' + req.get('host') + '/rest/2.2/assets/user/' + testAssetId;

        var form = new multiparty.Form();
        var hasAssetField = false;
        var hasTitleField = false;

        form.on('error', function() {
            errors = 'Server Error';
            res.status(500).end();
        });

        form.on('part', function(part) {
            // filename is "null" when this is a field and not a file
            if (!part.filename) {
                if (part.name === 'title') {
                    hasTitleField = true;
                }
            }

            if (part.filename && part.name) {
                if (part.name === 'asset') {
                    hasAssetField = true;
                }
            }

            // don't read any data
            part.resume();
        });

        form.on('close', function() {
            if (!hasAssetField || !hasTitleField) {
                return res.status(400).end('Bad request. Missing \'asset\' file field or title field');
            }

            res.status(201).location(assetUrl).end();
        });

        form.parse(req);
    });

    app.put('/rest/2.2/users/:user_id/mailings/:mailing_id/recipients', function(req, res) {

        var body = req.body;

        if (!body || !body.recipientFields || !body.recipients
            || body.recipients.length <= 0
            || body.recipients[0].length !== body.recipientFields.length
        ) {

            return res.status(400)
                .send('Bad Request, missing or bad JSON PUT `recipients` and `recipientFields` parameters')
                .end();

        }

        // all is fine, return a 204 No Content
        res.status(204).end();
    });

    var handleSVGPage = function (req, res) {
        if (req.get('Content-Type') !== 'image/svg+xml') {
            return res.status(400).end('Bad Content-Type');
        }

        if (!req.body.toString().indexOf('<svg')) {
            return res.status(400).end('No SVG data found');
        }

        // all is fine, return a 204 No Content
        res.status(204).end();
    }

    app.put('/rest/2.2/users/:user_id/mailings/:mailing_id/pages/1', handleSVGPage);
    app.put('/rest/2.2/users/:user_id/mailings/:mailing_id/pages/2', handleSVGPage);


    app.post('/rest/2.2/users/:user_id/mailings/:mailing_id/order', function (req, res) {
        var testOrderUrl = req.protocol + '://' + req.get('host') + '/rest/2.2/users' + req.params.user_id + '/orders/' + testOrderId;
        var body = req.body;
        /*
        {
            "paymentType": "SWISS_POST",
            "division": "POST_MAIL",
            "costCenter": "64000168",
            "invoiceRefNo": "50000024"
        }
        */

        if (!body || body.paymentType !== 'SWISS_POST' || body.division !== 'POST_MAIL' || !body.costCenter || body.invoiceRefNo !== '50000024') {
            return res.status(400).end('Bad request, missing JSON parameters');
        }

        res.status(201).location(testOrderUrl).end();
    });

    // Start server
    var server = http.createServer(app).listen(port);

    return server;
}

module.exports = createTestServer;


spec/users.spec.js
var postcardcreator = require('../index.js');
var stream = require('stream');

describe('Users', function() {
    var client = new postcardcreator.Client('ABC123456');

    var unauthenticatedClient = new postcardcreator.Client();

    beforeEach(function() {
        spyOn(client, 'request');
        spyOn(unauthenticatedClient, 'request')
    });

    it('get the current user', function() {
        client.users.current();

        expect(client.request).toHaveBeenCalledWith({
            url: '/users/current',
            method: 'GET'
        }, undefined);
    });

    it('get the quota user', function() {
        client.users.quota.get(456);

        expect(client.request).toHaveBeenCalledWith({
            url: '/users/456/quota',
            method: 'GET',
        }, undefined);
    });

    it('create a new user', function() {
        var user = {
            tenantId: "CHE",
            email: 'charles@example.org',
            password: 'passw0rd',
            company: null,
            sex: 'MALE',
            givenName: 'Charles',
            familyName: 'Dickens',
            language: 'en',
            newsletterSubscribed: false,
            gtcAccepted: true,
            address: "Rue du Roman 7",
            postCode: "1000",
            place: "Lausanne",
            country: "CHE"
        };

        unauthenticatedClient.users.create(user);

        expect(unauthenticatedClient.request).toHaveBeenCalledWith({
            url: '/users',
            method: 'POST',
            body: user,
            json: true
        }, undefined);
    });


    it('create the mailings', function() {
        var callback = function () {};
        client.users.mailings.create(12345, callback);

        expect(client.request).toHaveBeenCalled();

        expect(client.request.mostRecentCall.args[0].url).toBe('/users/12345/mailings');
        expect(client.request.mostRecentCall.args[0].method).toBe('POST');
        expect(client.request.mostRecentCall.args[0].body).toBeDefined();

        // name is generated based on current datetime
        expect(client.request.mostRecentCall.args[0].body.name).not.toBeNull();
        expect(client.request.mostRecentCall.args[0].body.productId).toBe(2);
        expect(client.request.mostRecentCall.args[0].body.source).toBe('MOBILE');
        expect(client.request.mostRecentCall.args[0].body.addressFormat).toBe('PERSON_FIRST');

        expect(client.request.mostRecentCall.args[1]).toBe(callback);
    });

    it('upload the asset', function() {
        // create a fake stream
        var asset = new stream.Readable();
        asset._read = function noop() {}; // this is needed for node v0.10.26
        asset.push('[dataimagehere]');

        client.users.assets.create(12345, {
            title: 'My asset title',
            asset: asset
        });

        expect(client.request).toHaveBeenCalledWith({
            url: '/users/12345/assets',
            method: 'POST',
            formData: {
                title: 'My asset title',
                asset: asset
            },
            json: true
        }, undefined);
    });

});


spec/client.spec.js
var postcardcreator = require('../index.js');

describe('Client', function() {

    it('should build the correct baseUrl', function() {
        var client = new postcardcreator.Client('ABC123456', {
            'host': 'google.com'
        });

        expect(client.getBaseUrl()).toBe('https://google.com/rest/2.2');
    });

    it('should build the correct baseUrl without SSL', function() {
        var client = new postcardcreator.Client('ABC123456', {
            'host': 'google.com',
            'useSSL': false,
        });

        expect(client.getBaseUrl()).toBe('http://google.com/rest/2.2');
    });
});


lib/Client.js
var request = require('request');
var Q = require('q');
var _ = require('underscore');
var Users = require('./Users');
var DefaultLayout = require('./Layout');

var defaultHost = 'postcardcreator.post.ch';

/**
 Client constructor.
 @constructor
 @param {string} authToken
 @param {object} options (optional)
 - @member {string} host
 - @member {bool} useSSL
 - @member {string} proxy
 */
function Client(authToken, options) {
    options = options || {};
    this.authToken = authToken;
    this.host = options.host || defaultHost;
    this.proxy = options.proxy || null;

    if (typeof options.useSSL !== "undefined") {
        this.useSSL = options.useSSL;
    } else {
        // default value
        this.useSSL = true;
    }

    var usersResources = Users(this);
    this.users = usersResources;
}

/**
 Send a Postcard.

 @param {Postcard} postcard object
 @param {function} callback (optional)
 */
Client.prototype.sendPostcard = function (postcard, callback) {
    var deferred = Q.defer();
    var client = this;

    callback = (typeof options === 'function' ? options : callback);
    options = (typeof options === 'object' ? options : {});

    var postcardResult = {
        postcard: postcard,
        recipient: postcard.recipient,
        message: postcard.message,
    };

    /*
     * @TODO argh! I don't how to cleanup this code by using promise
     * without loosing variable scope (userId, mailingId, assetId, ...)
     */
    client.users.current(function(err, user) {

        if (err) {
            console.log('** ERROR ** on users.current() ', err);
            return deferred.reject(err);
        }

        var userId = user.userId;
        postcardResult.user = user;

        // create the mailing
        client.users.mailings.create(userId, {}, function (err, result) {

            if (err) {
                console.log('** ERROR ** on mailings.create() ', err);
                return deferred.reject(err);
            }

            var mailingId = result.nodeClientResponse.headers['location'].match(/[0-9]+$/)[0];
            postcardResult.mailingId = mailingId;

            // upload assets
            client.users.assets.create(userId, { asset: postcard.image }, function (err, result) {

                if (err) {
                    console.log('** ERROR during assets.create() ', err);
                    return deferred.reject(err);
                }

                var assetId = result.nodeClientResponse.headers['location'].match(/[0-9]+$/)[0];
                postcardResult.assetId = assetId;

                // set recipients
                client.users.mailings.recipients.put(userId, mailingId, postcard.recipient, function(err, result) {

                    if (err) {
                        console.log('** ERROR during recipients.put() ** ', err);
                        return deferred.reject(err);
                    }

                    var sender = user;
                    sender.street = user.address; // to be more API compliant

                    var frontLayout = postcard.renderFrontPage({ assetId: assetId });
                    var backLayout = postcard.renderBackPage({ sender: sender });

                    // define the layout of the frontpage
                    client.users.mailings.pages.front.put(userId, mailingId, frontLayout, function(err, result) {

                        if (err) {
                            console.log('** ERROR during mailings.pages.front.put() ** ', err);
                            return deferred.reject(err);
                        }

                        // define the layout of the back page
                        client.users.mailings.pages.back.put(userId, mailingId, backLayout, function(err, result) {

                            if (err) {
                                console.log('** ERROR during mailings.pages.back.put() ** ', err);
                                return deferred.reject(err);
                            }

                            // Create the Order
                            client.users.mailings.order.create(userId, mailingId, {}, function(err, result) {

                                if (err) {
                                    console.log('** ERROR during mailings.order.create() ** ', err);
                                    return deferred.reject(err);
                                }

                                var orderId = result.nodeClientResponse.headers['location'].match(/[0-9]+$/)[0];
                                postcardResult.orderId = orderId;

                                // ** END **
                                deferred.resolve(postcardResult);
                            });
                        });
                    });
                });
            });
        });
    });

    return deferred.promise.nodeify(callback);
};

/**
 Get the base Url which will be used for all request with this client.

 @returns {string} - the API base URL
 */
Client.prototype.getBaseUrl = function () {

    return (this.useSSL ? 'https':'http') + '://' + this.host + '/rest/2.2';
};

/**
 Make an authenticated request against the API.

 @param {object} options - options for HTTP request.
 @param {function} callback - when request is complete
 - @param {object} error - an error object if there is a problem
 - @param {object} data - the JSON-parsed data
 */
Client.prototype.request = function (options, callback) {
    var deferred = Q.defer();
    var client = this;

    if (_.isString(options)) {
        options = { url: options, method: 'GET', json: true };
    }

    // Add base URL if we weren't given an absolute one
    if (options.url.indexOf('http') !== 0) {
        options.url = client.getBaseUrl() + options.url;
    }

    options.headers = options.headers || {};
    options.headers['Accept'] = '*/*';
    options.headers['Accept-Charset'] = 'utf-8';
    options.headers['User-Agent'] = 'Mozilla/5.0 (Linux; Android 5.1; Nexus 5 Build/LMY47I) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/36.0.1985.125 Mobile';

    if (this.authToken) {
      options.headers['Authorization'] = 'Bearer ' + this.authToken;
    }

    if (this.proxy) {
        options.proxy = this.proxy;
    }

    // do the request
    request(options, function(err, response, body) {
        var data;

        try {
            if (err) {
                data = err;
            } else {
                data = body ? JSON.parse(body) : {};
            }
        } catch (e) {
            data = { status: 500, message: (e.message || 'Invalid JSON body') };
        }

        //request doesn't think 4xx is an error - we want an error for any non-2xx status codes
        var error = null;
        if (err || (response && (response.statusCode < 200 || response.statusCode > 206))) {
            error = {};
            // response is null if server is unreachable
            if (response) {
                error.status = response.statusCode;
                error.message = 'Unable to complete HTTP request';
            } else {
                error.status = 500;
                error.message = 'Unable to reach host: "'+client.host+'"';
            }
        }

        //hang response off the JSON-serialized data, as unenumerable to allow for stringify.
        Object.defineProperty(data, 'nodeClientResponse', {
            value: response,
            configurable: true,
            writeable: true,
            enumerable: false
        });

        error && Object.defineProperty(error, 'nodeClientResponse', {
            value: response,
            configurable: true,
            writeable: true,
            enumerable: false
        });

        if (error) {
            deferred.reject(error);
        } else {
            deferred.resolve(data);
        }
    });


    // Return promise, but also support original node callback style
    return deferred.promise.nodeify(callback);
};

module.exports = Client;


lib/Layout.js
var _ = require('underscore'),
    fs = require('fs'),
    path = require('path');

var Layout = {};

Layout.FRONT_SVG_TEMPLATE_PATH = path.join(__dirname, 'template', 'front.svg.template');
Layout.BACK_SVG_TEMPLATE_PATH = path.join(__dirname, 'template', 'back.svg.template');

/*
 @var
  - assetId
*/
Layout.getFrontPage = function(data) {
    var tpl = fs.readFileSync(this.FRONT_SVG_TEMPLATE_PATH).toString();
    var template = _.template(tpl);

    return template(data);
};

/*
 @var
 - message
 - recipient.givenName
 - recipient.familyName
 - recipient.company
 - recipient.street
 - recipient.postCode
 - recipient.place
 - sender.givenName
 - sender.familyName
 - sender.company
 - sender.address  <-- /!\
 - sender.postCode
 - sender.place
*/
Layout.getBackPage = function (data) {
    var tpl = fs.readFileSync(this.BACK_SVG_TEMPLATE_PATH).toString();
    var template = _.template(tpl);

    return template(data);
};

module.exports = Layout;


lib/Postcard.js
var _ = require('underscore');
var fs = require('fs');
var DefaultLayout = require('./Layout');

/**
 Postcard object

 @param {stream.Readable} image
 @param {string} message
 @param {object} recipient the address of the person you want to send the postcard to
     - @member givenName
     - @member familyName
     - @member company
     - @member street
     - @member postCode
     - @member place
 @param {object} sender address of the postcard (optional) If not defined will be automatically filled with authenticated user.
     - @member givenName
     - @member familyName
     - @member company
     - @member street
     - @member postCode
     - @member place
 @param {object} options (optional)
 */
function Postcard(image, message, recipient, sender, options) {
    this.image = image;
    this.recipient = recipient;
    this.message = message;
    this.sender = sender;
    this.options = options || {};
};

/**
 Render the front page layout.

 @param data
    - assetId

 @return string of a SVG rendered template
*/
Postcard.prototype.renderFrontPage = function (data) {

    return DefaultLayout.getFrontPage(data);
};

/**
 Render the back page layout.

 @param missingData
    - sender

 @return string of a SVG rendered template
 */
Postcard.prototype.renderBackPage = function (missingData) {
    var data = {};

    data.message = this.message || missingData.message;
    data.recipient = this.recipient || missingData.recipient;
    data.sender = this.sender || missingData.sender;

    return DefaultLayout.getBackPage(data);
};

module.exports = Postcard;


lib/AddressBuilder.js
var AddressBuilder = {};

AddressBuilder.requiredFields = {
    'salutation': {
        "name": "Salutation",
        "addressField": "SALUTATION"
    },
    'givenName': {
        "name": "Given Name",
        "addressField": "GIVEN_NAME"
    },
    'familyName': {
        "name": "Family Name",
        "addressField": "FAMILY_NAME"
    },
    'company': {
        "name": "Company",
        "addressField": "COMPANY"
    },
    'companyAddition': {
        "name": "Company",
        "addressField": "COMPANY_ADDITION"
    },
    'street': {
        "name": "Street",
        "addressField": "STREET"
    },
    'postCode': {
        "name": "Post Code",
        "addressField": "ZIP_CODE"
    },
    'place': {
        "name": "Place",
        "addressField": "PLACE"
    }
};

/*
 * Build recipients JSON data before sending it to the API
 */
AddressBuilder.buildRecipients = function (recipient) {
    var recipientFields = [];
    var recipientsData = [];

    var keys = Object.keys(AddressBuilder.requiredFields);
    for (var i in keys) {
        var key = keys[i];
        recipientFields.push(AddressBuilder.requiredFields[key]);
        if (typeof recipient[key] !== "undefined") {
            recipientsData.push(recipient[key]);
        } else {
            recipientsData.push("");
        }
    }

    return {
        recipientFields: recipientFields,
        recipients: [ recipientsData ]
    };
};

module.exports = AddressBuilder;


examples/sendpostcard.js
var postcardcreator = require('../lib'),
    Postcard = postcardcreator.Postcard,
    fs = require('fs'),
    path = require('path'),

    postcard_user = process.env.POST_USER,
    postcard_pass = process.env.POST_PASSWORD,
    options = {};


if (process.env.TEST) {
    postcard_user = 'bob@example.org';
    postcard_pass = 'ilovealice';
    options = {
        host: 'localhost:3001',
        useSSL: false
    };
}

var client = new postcardcreator.Client(postcard_user, postcard_pass, options);

console.log("Created a client with credentials "+postcard_user);


// Example of using promise
var promise = client.users.current();

promise.then(function(user) {
    console.log('User ID', user.userId);
    console.log('User Email', user.email);
    console.log(user);

    // ... and here an exemple using callback
    client.users.quota.get(user.userId, function(err, quota) {

        if (err) {
            console.log('Error : ', err);
            return;
        }

        if (quota.quota <= 0 || quota.globalQuotaExceeded) {
            console.log('Quota exceeded, need to wait 24h :(');
            return;
        }

        console.log('Quota not exceeded :)');
    });

}, function (error) {
    console.log('ERREUR: ', error);
});


// Exemple of sending a PostCard
var message = "Ceci est un test."; // your message
var assetStream = fs.createReadStream(path.join(__dirname, 'images/yann.jpg')); // your image
var recipient = {
    salutation: "Monsieur",
    givenName: "Gilles",
    familyName: "Doge",
    company: "Antistatique.net",
    street: "Rue de Sébeillon 9b",
    postCode: "1004",
    place: "Lausanne"
};

var postcard = new Postcard(assetStream, message, recipient);

client.sendPostcard(postcard, function(err, result) {
    if (!err) {
        console.log("Postcard sent with success !");
    } else {
        console.log("Error when sending the postcard. ", err);
    }
});


lib/Users.js
var fs = require('fs');
var dateFormat = require('dateformat');
var AddressBuilder = require('./AddressBuilder');

module.exports = function (client) {

    function current(callback) {
        var requestParams = {
            url: '/users/current',
            method: 'GET'
        };

        return client.request(requestParams, callback);
    }

    function usersPost(params, callback) {
        var requestParams = {
            url: '/users',
            method: 'POST',
            body: params,
            json: true
        };

        return client.request(requestParams, callback);
    }

    function quota(userId, callback) {
        var requestParams = {
            url: '/users/' + userId + '/quota',
            method: 'GET',
        };

        return client.request(requestParams, callback);
    }

    function mailingsPost(userId, params, callback) {

        callback = (typeof params === 'function' ? params : callback);
        params = (typeof params === 'function' ? {} : params) || {};

        // set default value that just works
        params.name = params.name || 'Mobile App mailing ' + dateFormat(new Date(), 'yyyy-dd-mm HH:MM');
        params.productId = params.productId || 2;
        params.source = params.source || 'MOBILE';
        params.addressFormat = params.addressFormat || 'PERSON_FIRST';

        var requestParams = {
            url: '/users/' + userId + '/mailings',
            method: 'POST',
            body: params,
            json: true
        };

        return client.request(requestParams, callback);
    }

    function recipientsPut(userId, mailingId, params, callback) {

        var recipients = AddressBuilder.buildRecipients(params);
        var opts = {
            url: '/users/' + userId + '/mailings/' + mailingId + '/recipients',
            method: 'PUT',
            body: recipients,
            json: true
        };

        return client.request(opts, callback);
    }

    function assetsPost(userId, params, callback) {
        params = params || {};

        var formData = {
            title: params.title || "Title of image",
            asset: params.asset //fs.createReadStream(assetFilename)
        };

        var opts = {
            url: '/users/' + userId + '/assets',
            method: 'POST',
            formData: formData,
            json: true
        };

        return client.request(opts, callback);
    }

    function generatePagePut(pageNumber) {

        return function pagePut(userId, mailingId, page, callback) {
            var opts = {
                url: '/users/' + userId + '/mailings/' + mailingId + '/pages/' + pageNumber,
                method: 'PUT',
                body: page,
                headers: {
                    'Content-Type': 'image/svg+xml'
                }
            };

            return client.request(opts, callback);
        };
    }

    function orderPost(userId, mailingId, params, callback) {
        params = params || {};

        params.paymentType = params.paymentType || 'SWISS_POST';
        params.division = params.division || 'POST_MAIL';
        params.costCenter = params.costCenter || "64000168";
        params.invoiceRefNo = params.invoiceRefNo || "50000024";

        var opts = {
            url: '/users/' + userId + '/mailings/' + mailingId + '/order',
            method: 'POST',
            body: params,
            json: true
        };

        return client.request(opts, callback);
    }

    /**
     @TODO: find a way to not be forced to specify the userId in each API endpoint
     A nicer API whould be:
        client.users(1234).quota(); // return the quota of user id 1234
        client.users().quota();     // return the quota of the authenticated user
     */
    var resourceApi = {
        current: current,
        create: usersPost,
        post: usersPost,

        quota: {
            get: quota
        },

        mailings: {
            create: mailingsPost,
            post: mailingsPost,
            recipients: {
                put: recipientsPut,
            },
            pages: {
                front: {
                    put: generatePagePut(1)
                },
                back: {
                    put: generatePagePut(2)
                }
            },
            order: {
                create: orderPost,
                post: orderPost
            }
        },

        assets: {
            create: assetsPost,
            post: assetsPost
        }
    };

    return resourceApi;
};
